### PR TITLE
Fix t0090 tar&gz unexpected EOF error

### PR DESCRIPTION
**Pull request by rht**

### CODE REVIEW

**Comment by whyrusleeping** [comment]
could i get a tldr of what was wrong?

File: unixfs/archive/tar/writer.go [hunk 1]
```diff
@@ -60,8 +60,11 @@ func (w *Writer) writeFile(nd *mdag.Node, pb *upb.Data, fpath string) error {
 	}
 
 	dagr := uio.NewDataFileReader(w.ctx, nd, pb, w.Dag)
-	_, err := dagr.WriteTo(w.TarW)
-	return err
+	if _, err := dagr.WriteTo(w.TarW); err != nil {
+		return err
+	}
+	w.TarW.Flush()
```
Comment (rht) [inline unixfs/archive/tar/writer.go]: This one fixes the tar EOF err

File: unixfs/archive/archive.go [hunk 2]
```diff
@@ -53,6 +66,11 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
 				pipew.CloseWithError(err)
 				return
 			}
+			maybeGzw.Close()
```
Comment (rht) [inline unixfs/archive/archive.go]: explicit Close is needed for a gzip writer. Reading the source, it write the end record etc.

File: unixfs/archive/archive.go [hunk 3]
```diff
@@ -29,15 +41,16 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
 	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
 
 	// compression determines whether to use gzip compression.
-	var maybeGzw io.Writer
+	var maybeGzw io.WriteCloser
+	var err error
```
Comment (jbenet) [inline unixfs/archive/archive.go]: i thought this function (`DagArchive`) had gotten decomposed. maybe i misremember. it's really big and scary though...

**Comment by jbenet** [comment]
@rht great work fixing these!

@rht @whyrusleeping can you help decompose out this func? if not, i'll give it a shot in the next week. this can be a separate PR.

**Comment by jbenet** [comment]
LGTM. @whyrusleeping you?

**Comment by whyrusleeping** [comment]
LGTM, more tests pass than did before!

File: unixfs/archive/archive.go (hunk 3 above)
Comment (rht) [inline unixfs/archive/archive.go]: It was archivewriter -> DagArchive and tar.NewWriter

**Pull request merged**